### PR TITLE
simulator first version

### PR DESCRIPTION
**Pull request by ThomasDCepfl**
Dear all,
Here is a first version of the simulator I have been working on throughout my semester project at MLO.
I tried to make it as generic and easy-to-use as possible.
I split the different aspects into different packages, which all have their dedicated notebook, Python files and README. 
I am open to any suggestions or remarks.
Best,
Thomas

### CODE REVIEW

**Review by koloskova — CHANGES_REQUESTED** [review]
The code looks good and implements many useful features, however, the code structure could be improved. There are many repetitions in the code which are good to be avoided.

File: simulator/A-NC/A_NC.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from utils import *
```
Comment (IamTao) [inline simulator/A-NC/A_NC.py:1]: it is better to explicitly list the functions/classes names.

File: simulator/A-NC/A_NC.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    """
+    Runs a decentralized optimization algorithme for the given learning rate for
+    a number of rounds, over some network. Outputs the accuracies and returns them.
+
+    Params:
+        train_loader (array): the list of all train datasets, one per client
+        test_loader (array): the list of test datasets, one per client
+        comm_matrix (numpy.array): the communication matric modeling the network
+        num_rounds (int): the number of data exchanges between nodes
+        epochs (int): the number of optimization steps between each communication (minimum 1)
```
Comment (IamTao) [inline simulator/A-NC/A_NC.py:14]: the `epoch` here is confusing. is it correspond to the number of local update steps?

File: simulator/A-NC/A_NC.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    accs = []
+    global_model, client_models = model_init(num_clients, net)
+    opt = optimizer_init(client_models, lr, optimizer)
+
+    loss, test_loss, acc = 0.0, 0.0, 0.0
+    for r in range(num_rounds):
+        loss = 0
+        for i in range(num_clients):
+            loss += client_update(client_models[i], opt[i], train_loader[i], epoch=epochs)
+        diffuse_params(client_models, comm_matrix)
```
Comment (IamTao) [inline simulator/A-NC/A_NC.py:32]: maybe `diffuse_params` could be a class that can provide different choices of diffusion, e.g. with compressed communication?

File: simulator/A-NC/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
+# A-NC
```
Comment (IamTao) [inline simulator/A-NC/README.md:1]: it is a bit hard to infer the exact meaning of `A-NC`

File: simulator/A-NC/utils.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+    client_model.train()
+    for e in range(epoch):
+        for batch_idx, (data, target) in enumerate(train_loader):
+            data, target = data.cuda(), target.cuda()
+            optimizer.zero_grad()
+            output = client_model(data)
+            loss = F.nll_loss(output, target)
+            loss.backward()
+            optimizer.step()
+    return loss.item()
```
Comment (IamTao) [inline simulator/A-NC/utils.py:109]: in addition to the loss, we may need to track other metrics, e.g. top-1 acc?
also, right now, it seems like the function will only return the loss value of the last iterate. I think it is not sufficient?>

File: simulator/failures/utils.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    and average the consensus values of each parameter vector.
+
+    Params:
+        global_model (nn.Module): the global neural network averaging all the clients
+        client_models (array of Net): the list of all the client neural networks
+
+    Returns:
+        cons(array): the list of consensus values, one per client
+    """
+    cos = nn.CosineSimilarity(dim=0, eps=1e-6)
```
Comment (IamTao) [inline simulator/failures/utils.py:188]: why we initialize the `cos` here?

File: simulator/failures/utils.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        traindata = datasets.MNIST('./data', train=True, download=True,
+                       transform=transforms.Compose([transforms.ToTensor(),transforms.Normalize((0.1307,), (0.3081,))])
+                       )
+        traindata_split = torch.utils.data.random_split(traindata, [int(traindata.data.shape[0] / num_clients) for _ in range(num_clients)])
+        train_loader = [torch.utils.data.DataLoader(x, batch_size=batch_size, shuffle=True) for x in traindata_split]
+
+        test_loader = torch.utils.data.DataLoader(
+                datasets.MNIST('./data', train=False, transform=transforms.Compose([transforms.ToTensor(),transforms.Normalize((0.1307,), (0.3081,))])
+                ), batch_size=batch_size, shuffle=True)
+    else: # distribution == 'non-iid'
```
Comment (IamTao) [inline simulator/failures/utils.py:225]: we may include more description details and other degrees of non-iid-ness for the `non-iid` data distribution.

File: simulator/A-NC/utils.py [hunk 8]
```diff
@@ -0,0 +1,344 @@
+import numpy as np
+import networkx as nx
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+from tqdm import tqdm
+
+class Net(nn.Module):
```
[the remark this answers is not in the document]
Comment (IamTao) [reply simulator/A-NC/utils.py:10]: agreed

**Comment by IamTao** [comment]
the simulator code looks good to me. 

my main concern is that, why not build a unified simulation code (i.e., one package) that can cover different cases, e.g., topologies, failures, latency?
